nixos/quassel: refactor quassel service

maxaudron opened the pull request:
## Description of changes

refactor the quassel service to allow full declarative configuration and add new configuration options.

Database and Authentication settings can now be set declarative when the configFromEnvironment variable is set to true. This defaults to false as to prevent breaking existing configuration.

Also moved all quassel settings options into the settings block and rename and organize them according to quassels flag naming.


## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

maxaudron committed:
nixos/quassel: refactor quassel service

refactor the quassel service to allow full declarative configuration and
add new configuration options.

Database and Authentication settings can now be set declarative when
the configFromEnvironment variable is set to true. This defaults to
false as to prevent breaking existing configuration.

Also moved all quassel settings options into the settings block and
rename and organize them according to quassels flag naming.

h7x4 commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
-        description = lib.mdDoc ''
-          The directory holding configuration files, the SQlite database and the SSL Cert.
+          This can be used to securely store tokens and secrets outside of the world-readable Nix store.
+          Since this file is read by systemd, it may have permission 0400 and be owned by root.
```
I think this should be a note

```suggestion

          ::: {.note}
          Since this file is read by systemd, it may have permission `0400` and be owned by root.
          :::
```

h7x4 commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
-        '';
-      };
+      environmentFile = mkOption {
+        type = types.nullOr types.str;
```
```suggestion
        type = types.nullOr types.path;
```

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
+                            Specifies the LDAP authenticator bind password. Warning: do not set
+                            confidential information here because it is world-readable in the Nix store.
+                          '';
+                        };
```
These should be set through `environmentFile`, i've added that to the description

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
+                Configure quassels authenticator and database settings using environment variables,
+                Instead of imperatively setting it up using the setup wizard during first connection to the quassel core.
+              '';
+            };
```
The only options this effects are the db and auth sections, so these are already kinda separated, however as soon as `--config-from-environment` is active both the db backend and auth backend need to be set in the environment variables or the core errors on startup.
Also not sure of how to best enable it based on changes, as the core does not provide defaults for these values. we could introduce an enable options in both db and auth, and enable the flag according to that, but as stated, enabling one, would enable both. Potentially unintentionally disabling whatever configs the user might have configured "manually" via the core setup.

maxaudron committed:
nixos/quassel: use DynamicUser and remove quassel user

run the service using systemd's DynamicUser and change the default
dataDir for quassel to be the systemd created StateDirectory
`/var/lib/quassel`

Also removed the quassel user and group and associated IDs

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
-          The interfaces the Quassel daemon will be listening to.  If `[ 127.0.0.1 ]`,
-          only clients on the local host can connect to it; if `[ 0.0.0.0 ]`, clients
-          can access it from any network interface.
+      user = mkOption {
```
I've switched it over to using dataDir in a separate commit keeping the functionality of setting a custom dataDir and user intact.

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
+              type = types.enum [ "Debug" "Info" "Warning" "Error" ];
+              default = "Info";
+              description = ''
+                Supports one of Debug|Info|Warning|Error;
```
Changed the description to a generated one. There is sadly no information on what the log levels exactly do. And the previous description was pulled straight out of the --help :/

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
+
+                      TCP and UNIX Sockets are supported
+                    '';
+                    default = null;
```
I've kept the namespacing and removed the nulling, which was not needed anyways, slipped in from a previous attempt

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -111,29 +428,74 @@ in
-                                     ++ optional config.services.mysql.enable "mysql.service";
+        after = [ "network.target" ]
+          ++ optional config.services.postgresql.enable "postgresql.service"
+          ++ optional config.services.mysql.enable "mysql.service";
```
I've removed the afters and added notes to the postgresql and ldap sections about the user adding them

maxaudron commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -10,89 +10,406 @@ let
+                            Specifies the Postgres connection user password. Warning: do not set confidential
+                            information here because it is world-readable in the Nix store.
+                          '';
+                        };
```
Same as bindPassword, these should be set through environmentFile, i've added that to the description

maxaudron commented:
@h7x4 thanks for the review! I've addressed most points and think this is ready for a re-review.

h7x4 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -365,6 +365,20 @@
 
 - The `prayer` package as well as `services.prayer` have been removed because it's been unmaintained for several years and the author's website has vanished.
 
+- The `services.quassel` module has been refactored and extended to provide more
```
These would have to be moved to rl2405

h7x4 commented on nixos/modules/services/networking/quassel.nix:
```diff
@@ -6,134 +6,498 @@ let
-      user = mkOption {
-        default = null;
-        type = types.nullOr types.str;
+      settings = mkOption {
```
Considering this isn't a freeform module, I think it's a bit overkill to use submodules for every single layer of nesting. It would be a lot cleaner and easier to deal with, if this was just a big attrset of options.

Could you rewrite it to work like this?

```nix
      settings = {
        listen = mkOption {
          type = types.listOf types.str;
          default = [ "127.0.0.1" "::1" ];
          description = lib.mdDoc ''
            The address(es) quasselcore will listen on.
          '';
        };
        ...
        ident = {
          enable = lib.mkEnableOption (lib.mdDoc "internal ident daemon.");
          ...
        };
        ...
      };
```

h7x4 reviewed (commented):
Hello!

Sorry this has taken so much time. Here's a few more comments.
